scipy/integrate/_rules/_base.py

```diff
@@ -0,0 +1,657 @@
+        estimate_error(f, a, b) := \|higher.estimate(f, a, b) - lower.estimate(f, a, b)|
+
+    (where the absolute value is taken elementwise).
+
+    Attributes
+    ----------
+    higher : Rule
+        Higher accuracy rule.
+
+    lower : Rule
+        Lower accuracy rule.
+
+    See Also
+    --------
+    GaussKronrodQuad, NestedRule
+
+    Examples
+    --------
+
+    >>> from scipy.integrate import cub
+    >>> from scipy.integrate._rules import (
+    ...     GaussLegendreQuad, NestedFixedRule, ProductNestedFixed
+    ... )
+    >>> higher = GaussLegendreQuad(10)
+    >>> lower = GaussLegendreQuad(5)
+    >>> rule = NestedFixedRule(
+    ...     higher,
+    ...     lower
+    ... )
+    >>> rule_2d = ProductNestedFixed([rule, rule])
+    """
+
+    def __init__(self, higher, lower):
+        self.higher = higher
+        self.lower = lower
+
+    @property
+    def nodes_and_weights(self):
+        if self.higher is not None:
+            return self.higher.nodes_and_weights
+        else:
+            raise NotImplementedError
+
+    @property
+    def lower_nodes_and_weights(self):
+        if self.lower is not None:
+            return self.lower.nodes_and_weights
+        else:
+            raise NotImplementedError
+
+    def estimate_error(self, f, a, b, args=(), kwargs=None):
+        r"""
+        Estimate the error of the approximation for the integral of ``f`` in rectangular
+        region described by corners ``a`` and ``b``.
+
+        Parameters
+        ----------
+        f : callable
+            Function to estimate error for. ``f`` must have the signature::
+                f(x : ndarray, \*args, \*\*kwargs) -> ndarray
+
+            ``f`` should accepts arrays ``x`` of shape::
+                (npoints, ndim)
+
+            and output arrays of shape::
+                (npoints, output_dim_1, ..., output_dim_n)
+
+            In this case, ``estimate`` will return arrays of shape::
+                (output_dim_1, ..., output_dim_n)
+        a, b : ndarray
+            Lower and upper limits of integration as rank-1 arrays specifying the left
+            and right endpoints of the intervals being integrated over. Infinite limits
+            are currently not supported.
+        args : tuple, optional
+            Additional positional args passed to ``f``, if any.
+        kwargs : tuple, optional
+            Additional keyword args passed to ``f``, if any.
+
+        Returns
+        -------
+        err_est : ndarray
+            Result of error estimation. If ``f`` returns arrays of shape
+            ``(npoints, output_dim_1, ..., output_dim_n)``, then ``est`` will be
+            of shape ``(output_dim_1, ..., output_dim_n)``.
+        """
+
+        nodes, weights = self.nodes_and_weights
+        lower_nodes, lower_weights = self.lower_nodes_and_weights
+
+        error_nodes = np.concat([nodes, lower_nodes], axis=0)
+        error_weights = np.concat([weights, -lower_weights], axis=0)
```

Comment:
we'll have to use `np.concatenate` for now to support NumPy<2.0 (until array API standard support is added)